src/query/workerQueue.js

```diff
@@ -0,0 +1,238 @@
+    this.log = log
+
+    this.concurrency = c.ALPHA
+    this.queue = this.setupQueue()
+  }
+
+  /**
+   * Create the underlying async queue.
+   *
+   * @returns {Object}
+   */
+  setupQueue () {
+    const q = queue(this.processNext.bind(this), this.concurrency)
+
+    // If there's an error, stop the worker
+    q.error = (err) => {
+      this.log.error('queue', err)
+      this.stop(err)
+    }
+
+    // When all peers in the queue have been processed, stop the worker
+    q.drain = () => {
+      this.log('queue:drain')
+      this.stop()
+    }
+
+    // When a space opens up in the queue, add some more peers
+    q.unsaturated = () => {
+      if (this.running) {
+        this.log('queue:unsaturated')
+        this.fill()
+      }
+    }
+
+    q.buffer = 0
+
+    return q
+  }
+
+  /**
+   * Stop the worker, optionally providing an error to pass to the worker's
+   * callback.
+   *
+   * @param {Error} err
+   */
+  stop (err) {
+    if (!this.running) {
+      return
+    }
+
+    this.running = false
+    this.queue.kill()
+    this.log('worker:stop, %d workers still running', this.run.workers.filter(w => w.running).length)
+    this.callbackFn(err)
+  }
+
+  /**
+   * Use the queue from async to keep `concurrency` amount items running
+   * per path.
+   *
+   * @param {function(Error)} callback
+   */
+  execute (callback) {
+    this.running = true
+    this.callbackFn = callback
+    this.fill()
+  }
+
+  /**
+   * Add peers to the worker queue until there are enough to satisfy the
+   * worker queue concurrency.
+   * Note that we don't want to take any more than those required to satisfy
+   * concurrency from the peers-to-query queue, because we always want to
+   * query the closest peers to the key first, and new peers are continously
+   * being added to the peers-to-query queue.
+   */
+  fill () {
+    this.log('queue:fill')
+
+    // Note:
+    // - queue.running(): number of items that are currently running
+    // - queue.length(): the number of items that are waiting to be run
+    while (this.queue.running() + this.queue.length() < this.concurrency &&
+           this.path.peersToQuery.length > 0) {
+      this.queue.push(this.path.peersToQuery.dequeue())
+    }
+  }
+
+  /**
+   * Process the next peer in the queue
+   *
+   * @param {PeerId} peer
+   * @param {function(Error)} cb
+   * @returns {void}
+   */
+  processNext (peer, cb) {
+    if (!this.running) {
+      return cb()
+    }
+
+    // The paths must be disjoint, meaning that no two paths in the Query may
+    // traverse the same peer
+    if (this.run.peersSeen.has(peer)) {
+      return cb()
+    }
+
+    // Check if we've queried enough peers already
+    this.run.continueQuerying(this, (err, continueQuerying) => {
+      if (!this.running) {
+        return cb()
+      }
+
+      if (err) {
+        return cb(err)
+      }
+
+      // If we've queried enough peers, stop the queue
+      if (!continueQuerying) {
+        this.stop()
```

Comment:
Adding `this.stop` here improves the speed reliability quite a bit. Total query times are usually around or under 80seconds. While this isn't great, it's better than it currently is. The side effect of this, is that we might lose closer peers from the other, parallel, responses in this queue since it runs at a concurrency of 3.

I lowered the dial timeout (to 5s) on Switch for the js-ipfs instance I am running locally when testing against the live network. This also contributed to the improved speed.

One thing the Kademlia papers aren't particularly clear on (I may have missed it) is the behavior of the Disjoint Paths in terms of completing the query. Currently, we wait for all paths to finish before the query is finished. Should we? The more paths we allow to finish the closer we get to the **actual** closest peers, but the query takes significantly longer. If we stop after a single Disjoint Path has completed, the query time is dramatically reduced, but it's also less accurate. Long running peers might mitigated the accuracy problem as its routing table improves.

The big problem I am seeing now is slow/dead peers. My assumption is that the slow paths are hitting nodes that either aren't pruning their routing tables effectively, or the network is just that much in flux (nodes hitting the network and then going offline regularly). To remediate this we're either going to have to get really aggressive with the time we're allowing for connections and queries, or we'll have to be sloppier with what constitutes "closet". If we want to tighten down timeouts more, I think we'll want to add timeout options to `Switch.dial`. I've been changing this globally for testing, but DHT expectations vs global dial expectations are probably different.

A note: I played around with a concurrency of 6 instead of 3, but I actually observed a negative impact on query times.

@dirkmc @vasco-santos thoughts? I'm tempted to try things with the "sloppy" and fast approach and see how fetching performs on the network with that.

Reply:
Excellent work @jacobheun, these are solid improvements, taking orders of magnitude off the query time.

I agree that we should add timeout options to `Switch.dial()`. I think the main reason that we see so many slow/dead peers is because nodes that are behind NATs are advertising their peer ids to the network. I believe this is the main problem in the go DHT and a strong impetus for using hole punching techniques. Ideally nodes would have a mechanism to be able to check if they're behind a NAT before advertising their peer ids, and only do so if they are reachable through a relay.

I believe it's correct to wait for all disjoint paths to finish before returning. A streaming API should mitigate this problem.

I would be interested in seeing how well the sloppy approach works, but it sounds like it would need to be run in a simulated network to understand how it behaves in the wild (unless there is already some research out there?).

Reply:
Something I realized, which is likely attributing to some of the workers getting "blocked", are peer addresses. If we encounter a peer that is offline, and/or has a lot of bad addresses associated with it, the Switch timeout applies to per dial attempt, not connection. If we used the default dial timeout of 10seconds, and the peer had 20 unique addresses, it could take us 200 seconds to go through all of the addresses. We don't currently do parallel dials in the Switch on the same transport, as we can't yet abort dials properly if a connection succeeds. We will probably need to mitigate this with a temporary timeout on the entire connection attempt until the async iterators work is completed for transports, which will include abort logic so we can go back to using parallel dials.

Reply:
> because nodes that are behind NATs are advertising their peer ids to the network

Agreed, the NAT manager is on my list for this quarter, I may look at reprioritizing things a bit as this will help with performance in general.

I will look at adding an option to timeout the entire `Switch.dial`. I think this should help lower the high end of the queries.

If we go with the sloppy worker stop approach, and allow all disjoint paths to finish, we should still see reasonably close results. From https://github.com/libp2p/research-dht/issues/6, we could look at some other techniques to supplement the sloppy approach: 
> On a successful get, we re-publish the value to nodes on the lookup path. Subsequent gets to hot keys will find the result without. This is part of the original Kademlia algorithm that doesn't appear to be implemented.
> Sloppy provides. Analogous to re-publishing along the lookup path on get, provide doesn't need to go all the way to the "correct" node. If there are a lot of providers for a given key, new providers should be added to nodes on the lookup path. This is one of the fundamental innovations in Coral[3].

Reply:
> Total query times are usually around or under 80seconds

Not ideal, but a great improvement already! 👏 

> If we want to tighten down timeouts more, I think we'll want to add timeout options to Switch.dial. 

I believe this makes perfect sense for the DHT dials. As we will end up connecting with so many peers in the wild, the probability of failure will be super high and we should get to the next peer fast. So, I am totally in favor of this.

Regarding the disjoint paths, I also agree with @dirkmc that ideally, we should wait for all of them. However, we can make some experiments with that change, but I think it should be in a new PR, after this one. 

The sloppy approach can also be interesting to experiment, do you prefer to continue it here?

Reply:
I added the timeout to the query function. It seemed unnecessary to go about changing the Switch `dial` interface at this point. Once we have parallel dial aborts working, that should allow is to potentially remove the timeout here. During the Switch async refactor it might make sense to add the option to override timeouts per dial then.

> The sloppy approach can also be interesting to experiment, do you prefer to continue it here?

Right now this is sloppy. We're still running all disjoint paths, but we let workers end earlier than they might normally. One of the new tests validates the possibility of missing closer peers.


I want to look at adding a basic simulation test that we can look at running regularly when we attempt performance updates. But other than that, I think this is in a state worth trying out, and the next improvements will likely come from Switch and NAT improvements.